### ISSUE TITLE
Random tile of tilemap not rendering with canvas/WebGL when rotated

### ISSUE THREAD

**Issue by r3ndd** [craftyjs/Crafty#1200]
I have the following code for my scene:

```
var tiles;

var addTile = function (type, x, y, rot, tile_size) {
    var tile = Crafty.e("Tile").attr({x: x * tile_size, y: y * tile_size, rotation: rot || 0});
    tile.type(type);
    tile.origin("center");
    tiles.push(tile);
    if (type == "dirt_water")
        console.log(tile);
};

var addTileRows = function (type, startY, endY, width, rot, tile_size) {
    for (var y = startY; y < endY; y++) {
        for (var x = 0; x < width; x++) {
            addTile(type, x, y, rot, tile_size);
        }
    }
}

Crafty.scene("Main", function (data) {
    var width = 30;
    tiles = [];

    addTileRows("water", 0, 2, width, 0, data.tile_size);
    addTileRows("grass_water", 2, 3, width, -90, data.tile_size);
    addTileRows("grass", 3, 5, width, 0, data.tile_size);
    addTileRows("dirt_grass", 5, 6, width, -90, data.tile_size);
    addTileRows("dirt", 6, 10, width, 0, data.tile_size);
    addTileRows("dirt_water", 10, 11, width, 90, data.tile_size);
    addTileRows("water", 11, 12, width, 0, data.tile_size);
    addTileRows("sand_water", 12, 13, width, -90, data.tile_size);
    addTileRows("sand", 13, 16, width, 0, data.tile_size);
});
```

The tile component:
```
Crafty.c("Tile", {
    required: "2D, WebGL",

    init: function () {
        this.size = window.tile_size;
    },

    type: function (tile_name) {
        this.addComponent("spr_" + tile_name);
    }
});
```

However, I am getting an issue where one of the rotated tiles is not rendered: https://i.gyazo.com/9a526fd6febb991011f32043ca73fdce.png

Only happens on canvas and WebGL, DOM works fine. Adding 1 pixel to the x position of the tile makes the issue disappear, but that obviously isn't a viable fix for a game.

**Comment by r3ndd** [comment]
I should add that this is version 0.9.0, but after checking it appears 0.8.0 has the same issue.

**Comment by starwed** [comment]
Hmm, definitely seems like a bug.  If it occurs in WebGL *and* Canvas layers, but not DOM, and only for rotated entities, I bet it's an error determining whether the entity falls within the viewport.

I could imagine it being some sort of rounding error due to the rotation by `-90`, but in that case I'm not sure why it would only happen in a single row, instead of all of them.

Do you have a fiddle/etc that reproduces the behavior?  What environment (browser+OS) are you seeing it on?

**Comment by starwed** [comment]
It's possible that this will be fixed by the changes in #1202 -- changing the origin of an already-rotated entity was not updating some calculated properties correctly.

In particular, the position of the entity in the broad phase map was not updating correctly, and that's used for selecting which entities to render.  (For Canvas and WebGL, but not DOM layers.)

**Comment by starwed** [comment]
I've reproduced the issue in the `release` branch, and confirmed it is now fixed in the develop branch.  So it does seem like setting the origin after the rotation was the source of the problem -- not sure why that *specific* row was affected, but likely related to some quirk of how the rotated coordinates are rounded.

You can grab the most recent develop build of Crafty [here](https://github.com/craftyjs/Crafty-Distro).  A workaround when using 0.9 would be to specify the origin before setting the rotation property.

Thanks for the report!
